moderatorFrontend/src/main/kotlin/app/App.kt

```diff
@@ -126,40 +120,23 @@ private class App : RComponent<AppProps, AppState>() {
       }
     }
 
-  private fun enableClientSideRouting() {
-    window.addEventListener(PopStateEvent.POP_STATE, {
-      handleHistoryChange()
-    })
-
-    window.addEventListener(MouseEvent.CLICK, { event ->
-      val target = event.target
-      if (target != null && !event.altKey && !event.ctrlKey && !event.metaKey && !event.shiftKey) {
-        // Only handle click for anchor elements
-        val linkNode = (target as Node).findParent { it.nodeName.lowercase() == "a" } ?: return@addEventListener
-        val anchor = linkNode as web.html.HTMLAnchorElement
-        val parsedUrl = try {
-          URL(anchor.href)
-        } catch (_: Throwable) {
-          null
-        }
-        val relativeUrl = parsedUrl?.relativeUrl
-        val activeView = parsedUrl?.toRoute()
-        if (relativeUrl != null && anchor.target != WindowTarget._blank && activeView != null) {
-          if (relativeUrl != location.relativeUrl) {
-            history.pushState(null, anchor.title, relativeUrl)
-            handleHistoryChange()
-          }
-          event.preventDefault()
-        }
-      }
-    })
-  }
-
   private fun pushAppRoute(route: AppRoute) {
-    if (route.relativeUrl != location.relativeUrl) {
-      history.pushState(data = null, unused = route.url.title.get(), url = route.relativeUrl)
+    if (NavigationHandler.shouldNavigate(route, NavigationHandler.NavigationEvent.PUSH_APP_ROUTE, ::handleHistoryChange)) {
+      NavigationHandler.pushHistory(
+        relativeUrl = route.relativeUrl,
+        title = route.url.title.get(),
+        handleHistoryChange = ::handleHistoryChange,
+      )
+      // Reset scroll position.
+      // Otherwise, scroll position from previous route is kept.
+      window.scrollTo(
+        options = jso {
+          top = 0.0
+          left = 0.0
+          behavior = ScrollBehavior.instant
+        }
+      )
```

Comment:
Just a note, this scroll handling might cause the browser to not keep the scroll position when navigating back. This would be jarring - have you tested it?

Reply:
It would be a problem for `NavigationEvent.POPSTATE` (triggered on browser history backward/forward/go) but this is the `NavigationEvent.PUSH_APP_ROUTE` for new routes :)